pkg/controller/oss/oss_controller.go

```diff
@@ -84,41 +76,26 @@ func (c *Connector) Connect(ctx context.Context, mg resource.Managed) (managed.E
 		return nil, errors.New(errNotBucket)
 	}
 
-	var (
-		secretKeySelector *xpv1.SecretKeySelector
-		region            string
-	)
-
 	if err := c.Usage.Track(ctx, mg); err != nil {
 		return nil, errors.Wrap(err, errTrackUsage)
 	}
 
-	providerConfig, err := util.GetProviderConfig(ctx, c.Client, cr.Spec.ProviderConfigReference.Name)
+	pcName := cr.Spec.ProviderConfigReference.Name
+	cred, err := util.GetCredentials(ctx, c.Client, pcName)
 	if err != nil {
 		return nil, err
 	}
-
-	if s := providerConfig.Spec.Credentials.Source; s != xpv1.CredentialsSourceSecret {
-		return nil, errors.Errorf(errFmtUnsupportedCredSource, s)
-	}
-	secretKeySelector = providerConfig.Spec.Credentials.SecretRef
-	region = providerConfig.Spec.Region
-
-	if secretKeySelector == nil {
-		return nil, errors.New(errNoConnectionSecret)
-	}
-	s := &corev1.Secret{}
-	nn := types.NamespacedName{Namespace: secretKeySelector.Namespace, Name: secretKeySelector.Name}
-	if err := c.Client.Get(ctx, nn, s); err != nil {
-		return nil, errors.Wrap(err, errGetConnectionSecret)
+	region, err := util.GetRegion(ctx, c.Client, pcName)
+	if err != nil {
+		return nil, err
 	}
 
 	endpoint, err := util.GetEndpoint(cr.DeepCopyObject(), region)
```

Comment:
Why do we deep-copy `cr` here? Maybe:
```suggestion
	endpoint, err := util.GetEndpoint(cr, region)
```

Reply:
Then we can convert different kinds of managed resources to `runtime.Object` to check the type and use a common function to get endpoint for all cloud resources.

Reply:
In order to be able to call `util.GetEndpoint`, which expects a `runtime.Object` as its first argument if I recall correctly, you don't need to deep-copy it as `cr` already has the `runtime.Object` interface.

Reply:
You are absolutely right. I've fixed it in all managed resource controllers. Thanks.